### PR TITLE
Adding comments to .proto files

### PR DESCRIPTION
**Pull request by curino**
Initial round of `.proto` comments related to: Issue #215.

### CODE REVIEW

**Comment by jacques-n** [comment]
Is there a way to do comments to maximize the quality of the buf autogenerated documentation? Also, would be good to have a comment style checker. For example, I saw at least one place where your double slashes were not followed by a space.

**Comment by jvanstraten** [comment]
> Is there a way to do comments to maximize the quality of the buf autogenerated documentation?

I *think* it might support commonmark (i.e. disambiguated markdown), but I can't find any command or anything in buf's documentation about generating docs locally to play around with? What the? Am I just blind? I hope I'm blind, because having to publish to see what your docs looks like seems like an abysmal workflow.

If that's really just not a thing, could we use something like https://github.com/pseudomuto/protoc-gen-doc instead/in addition? IMO even better than just inlining proto3 code blocks with comments into the website and/or linking to the buf schema registry would be if the formatted part of the website is generated from them, making the code blocks mostly obsolete (hidden under a details block maybe, or behind a link a la sphinx and rustdoc). protoc-gen-doc can emit what looks to be a pretty simple JSON format with all the information relevant to documentation, so I'm sure I could whip up some Python script to convert that to something that integrates nicely with mkdocs and lints everything. Including a lint that ensures that every message actually appears on the website somewhere, so CI will force people to document their stuff when they add something.

> I saw at least one place where your double slashes were not followed by a space.

Likewise for starting comments in lowercase, not ending with a period, and line wrap point. I personally always use sentence case and punctuation and then manually wrap the comments at column 80. Makes them much easier to read than the one-all-lowercase-sentence-without-period-per-line style IMO.

**Comment by curino** [comment]
It would be great to have both some standard/guidelines on the comments (I was mostly winging it per various threads and common sense interpretation), as well as the ability to generate `buf` or similar docs and check them out. 

As for the auto-website generation it would be nice, but I am also ok for us to go over the existing website and inline references. In the long run the auto thingy you mentioned would be good. 

@jacques-n  let me know if there is anything you want me to address (other than I guess a check protobuf style). Also it would be nice to be able to trigger myself the checks (instead of waiting for someone to click go, so I can at least iterate on these things on my own before you need to look at it and review).

**Comment by jacques-n** [comment]
Hey @curino , on the checks, once we merge one of your patches, they will automatically run for future PRs. The manual run approval only applies to the very first PR. (or at least until one is merged by the PR proposer)

File: proto/substrait/algebra.proto [hunk 1]
```diff
@@ -13,16 +13,22 @@ option go_package = "github.com/substrait-io/substrait-go/proto";
...
+    // The underlying relation is output as is (no reordering or projection of columns)
     Direct direct = 1;
+    // Allows to control for order and inclusion of fields
     Emit emit = 2;
   }
 
   Hint hint = 3;
   substrait.extensions.AdvancedExtension advanced_extension = 4;
 
+  //Direct indicates no-op on ordering or presence of fields in the output
```
Comment (jacques-n) [inline proto/substrait/algebra.proto:28]: no-op seems weird here. No change or something similar might be a better way to say this.

File: proto/substrait/algebra.proto [hunk 2]
```diff
@@ -304,6 +339,7 @@ message RelRoot {
   repeated string names = 2;
 }
 
+// A relation without output field names (used internally in a plan)
```
Comment (jacques-n) [inline proto/substrait/algebra.proto:342]: The "without output field names" is strange. Let's state what it is, not what it isn't.

File: site/docs/relations/logical_relations.md [hunk 3]
```diff
@@ -13,6 +13,8 @@ The read operator is an operator that produces one output. A simple example woul
 | Property Maintenance | N/A (no inputs)                                              |
 | Direct Output Order  | Defaults to the schema of the data read after the optional projection (masked complex expression) is applied. |
 
+
```
Comment (jacques-n) [inline site/docs/relations/logical_relations.md:16]: Accidental modification?

**Review by jacques-n — COMMENTED** [review]
Mostly looks good. Did you confirm the layout for the site looks good for your embedded modifications? I suggest running `mkdocs serve` locally to make sure everything looks good.

File: site/docs/relations/logical_relations.md (hunk 3 above)
Comment (curino) [reply site/docs/relations/logical_relations.md:16]: yep

File: proto/substrait/algebra.proto (hunk 2 above)
Comment (curino) [reply proto/substrait/algebra.proto:342]: I agree it is odd to the specify what is not, but that is a bit suprising/unique for me, so something I felt like calling out. Happy to go either way if you feel strongly.

File: proto/substrait/algebra.proto (hunk 1 above)
Comment (curino) [reply proto/substrait/algebra.proto:28]: ACK

**Comment by curino** [comment]
> Mostly looks good. Did you confirm the layout for the site looks good for your embedded modifications? I suggest running `mkdocs serve` locally to make sure everything looks good.

Yes, I did check that and it looked good/consistent with the other. It is a basic embedding for now. I find it a bit redundant between the current text form (mostly specifiying fields) and the message, but we can get more nuanced later, for now it is a propagating/making things uniform.

File: proto/substrait/algebra.proto (hunk 2 above)
Comment (curino) [reply proto/substrait/algebra.proto:342]: ACK

**Comment by jacques-n** [comment]
Commit messages need to be fixed, otherwise looks good.

**Comment by curino** [comment]
> Commit messages need to be fixed, otherwise looks good.

Yeah, I tried but the force push didn't do what I wanted. I think I will create a separate clean branch with a single tidy commit (If I can guess what the style gods would like as a sacrifice ;-)).

**Comment by guykhazma** [comment]
@jacques-n sure I am happy to help. @curino please let me know if you would like me to help with finishing this PR.

**Comment by curino** [comment]
I created a new PR to have a clean commit history (hopefully) ;-)

**Pull request closed without merging**